include/swift/AST/DiagnosticsSema.def

```diff
@@ -2162,8 +2162,13 @@ NOTE(construct_raw_representable_from_unwrapped_value,none,
      "construct %0 from unwrapped %1 value", (Type, Type))
 
 // Derived conformances
-ERROR(cannot_synthesize_in_extension,none,
-      "implementation of %0 cannot be automatically synthesized in an extension", (Type))
+ERROR(cannot_synthesize_init_in_extension_of_nonfinal,none,
+      "implementation of %0 for non-final class cannot be automatically "
+      "synthesized in an extension due to 'init' requirement",
```

Comment:
One question here — excuse my mental lapse, but what is preventing us from synthesizing these? (I suspect that folks are going to hit this and be confused; we might want to be a bit more specific)

Reply:
In non-final classes, the `init`s satisfying protocol requirements must be `required` and those can't occur in extensions. I phrase the message this way because I couldn't get something more explanatory to be shorter, and it was pointed out to me that error messages aren't the place to teach the language. I'd love some alternative phrasings though, do you have any ideas?

Reply:
Ah, duh 🤦‍♂️ I haven't thought about it much, but I think cribbing some of the messaging we already have (e.g. the existing `"error: initializer requirement 'init(foo:)' can only be satisfied by a `required` initializer in the definition of non-final class 'Foo'"`) can help.

For instance, `"implementation of %0 for non-final class '%1' cannot be synthesized because initializer requirement 'required %2' cannot be implemented in an extension"` gets the user a little bit closer to understanding why this can't be done. Of course, the next question is "why does Swift treat a same-file extension any differently than writing things in the original declaration", but that's something for another day. 😅

Reply:
Yeah, that's a great idea! I've gone with:

> implementation of 'Decodable' for non-final class cannot be automatically synthesized in extension because initializer requirement 'init(from:)' can only be be satisfied by a 'required' initializer in the class definition